ballista/rust/core/proto/ballista.proto

```diff
@@ -151,6 +153,25 @@ message AggregateExprNode {
   LogicalExprNode expr = 2;
 }
 
+enum BuiltInWindowFunction {
+  ROW_NUMBER = 0;
+  RANK = 1;
+  DENSE_RANK = 2;
+  LAG = 3;
+  LEAD = 4;
+  FIRST_VALUE = 5;
+  LAST_VALUE = 6;
+}
+
+message WindowExprNode {
+  oneof window_function {
+    AggregateFunction aggr_function = 1;
```

Comment:
I checked whether this makes sense to reuse aggregate functions for window expressions - I think it does! E.g. PostgreSQL also says:
https://www.postgresql.org/docs/9.3/functions-window.html
> In addition to these functions, any built-in or user-defined aggregate function can be used as a window function (see Section 9.20 for a list of the built-in aggregates). Aggregate functions act as window functions only when an OVER clause follows the call; otherwise they act as regular aggregates.

Reply:
yes in general three types of things can be used:
1. aggregation
2. UDAF
3. built in window function

for both 1. and 2. they are not order sensitive, but for 3 we'll have to take sort into account

Reply:
```
[postgres] # explain select c1, count(c3) over (partition by c1 order by c3) from test;
                            QUERY PLAN
------------------------------------------------------------------
 WindowAgg  (cost=6.32..8.32 rows=100 width=12)
   ->  Sort  (cost=6.32..6.57 rows=100 width=4)
         Sort Key: c1, c3
         ->  Seq Scan on test  (cost=0.00..3.00 rows=100 width=4)
(4 rows)
```

```
[postgres] # explain select c1, first_value(c3) over (partition by c1 order by c3) from test;
                            QUERY PLAN
------------------------------------------------------------------
 WindowAgg  (cost=6.32..8.32 rows=100 width=6)
   ->  Sort  (cost=6.32..6.57 rows=100 width=4)
         Sort Key: c1, c3
         ->  Seq Scan on test  (cost=0.00..3.00 rows=100 width=4)
(4 rows)
```

IMO only the second time we'll really need to sort by `c3`

Reply:
also fun thing to notice:

```
[postgres] # explain analyze select c1, sum(c3) over (partition by c1 order by c3), avg(c3) over (partition by c1 order by c3 desc) from test;
                                                        QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------
 WindowAgg  (cost=11.64..13.64 rows=100 width=44) (actual time=1.287..1.373 rows=100 loops=1)
   ->  Sort  (cost=11.64..11.89 rows=100 width=36) (actual time=1.281..1.292 rows=100 loops=1)
         Sort Key: c1, c3
         Sort Method: quicksort  Memory: 31kB
         ->  WindowAgg  (cost=6.32..8.32 rows=100 width=36) (actual time=1.051..1.174 rows=100 loops=1)
               ->  Sort  (cost=6.32..6.57 rows=100 width=4) (actual time=0.221..0.231 rows=100 loops=1)
                     Sort Key: c1, c3 DESC
                     Sort Method: quicksort  Memory: 29kB
                     ->  Seq Scan on test  (cost=0.00..3.00 rows=100 width=4) (actual time=0.010..0.028 rows=100 loops=1)
 Planning Time: 0.087 ms
 Execution Time: 1.437 ms
(11 rows)
```

Reply:
> I checked whether this makes sense to reuse aggregate functions for window expressions - I think it does! E.g. PostgreSQL also says:
> https://www.postgresql.org/docs/9.3/functions-window.html
> 
> > In addition to these functions, any built-in or user-defined aggregate function can be used as a window function (see Section 9.20 for a list of the built-in aggregates). Aggregate functions act as window functions only when an OVER clause follows the call; otherwise they act as regular aggregates.

it is very useful for analytics, e.g. if you want to know (in an employee table with name, department, and salary) the list of employees in each department with salary above average.

Reply:
I believe `count(..) over order by ..` also needs to be sorted, it will do a count over the window, which means a running count by default.
But yeah very useful for analytics indeed :+1:

Reply:
My comment above was more about re-using the same functions over here - as I thought we might not want to support every aggregation function here too. But for me it sounds like a good idea to reuse them. Maybe @alamb has some ideas about it as well

Reply:
I think SQL is confusing in this area -- as @Jimexist  says, all "normal" aggregate functions (e.g. sum, count, etc) are also valid window functions, but the reverse is not true. You can't use window functions (e.g. LAG, LEAD, etc) outside of a window clause.

Thus I think representing window functions as a new type of function, as this PR does, makes the most sense. They are different enough (e.g. require information on the incoming windows) that trying to wrangle them into the same structures as normal aggregates seems like it will get messy. Long term I would expect we have a UDWF (user defined window function) api as well.

Ideally the physical implementation for `sum` / `count` / etc can be mostly reused but in the plans I think they are different enough to warrant different plan structures.

Reply:
> I believe count(..) over order by .. also needs to be sorted, it will do a count over the window, which means a running count (over sorted rows) by default.

Good point! Indeed.